### ISSUE TITLE
Best way to support you bringing this forward?

### ISSUE THREAD

**Issue by codiflow** [PatWie/excalidraw-complete#9]
Hey @PatWie, I really appreciate your work and now the builds from the `dev` channel are working. 🥳

But there's still some work to do until the firebase storage is being replaced completely.

One thing that still seems to be missing is the **storage of blobs** (images and other uploads). They don't use `firestore.googleapis.com` but `firebasestorage.googleapis.com` and those requests are not being replaced currently.

I'm a DevOps guy and not a developer but I really would like to help you with bringing this forward. What is the best way to help you with this?

Thanks 😎
Christian

**Comment by codiflow** [comment]
Now as I was playing around with your repo a bit these are my results:

My setup:
- I was replacing `locahost:3002` in `main.go` with a publicly resolvable hostname which is SSL terminated by nginx
- I replaced `"ssl=!0", "ssl=0"` with `"ssl=!0", "ssl=1"` and `"ssl:!0", "ssl:0"` with `"ssl:!0", "ssl:1"` to be able to use HTTPS
- I also set up a second setup using all the default values to cross-check the functionality

---

My findings:
1. **Collaboration** is working out of the box without issues ✅
2. The **backend storage is not working (correctly)** ❌ (neither in my SSL terminated setup nor in the default localhost setup). So you will always get the following error message after starting a collaboration session:
> Couldn't save to the backend database. If problems persist, you should save your file locally to ensure you don't lose your work.

3. Apart from that requests like the following are getting through ✅:
```
http://localhost:3002/v1/projects/excalidraw-room-persistence/databases/(default)/documents:batchGet
```

4. Using encryption functions in the browser is **only possible with SSL encrypted connections**. You will get errors in the browser console if thats the case. 💡

5. I recognized that the following **initial request is failing** in both variants (SSL terminated and localhost). This is one of the requests where `firebase.googleapis.com` was replaced by `localhost:3002`:
```
http://localhost:3002/google.firestore.v1.Firestore/Listen/channel?database=projects%2Fexcalidraw-room-persistence%2Fdatabases%2F(default)&VER=8&RID=40837&CVER=22&X-HTTP-Session-Id=gsessionid&%24httpHeaders=X-Goog-Api-Client%3Agl-js%2F%20fire%2F8.3.3%0D%0AX-Firebase-GMPID%3A1%3A654800341332%3Aweb%3A4a692de832b55bd57ce0c1%0D%0AContent-Type%3Atext%2Fplain%0D%0A&zx=66dvi15g5r94&t=1
```

6. The **firebase storage** `firebasestorage.googleapis.com` (which is being used for saving blobs like images) is **not being hot-patched yet**. You can try that on your own by uploading an image in Excalidraw and watch the browser logs (you have to be in a collaboration session during this test otherwise the image is only saved in the browsers storage) ❌

This is an example request:
```
OPTIONS https://firebasestorage.googleapis.com/v0/b//o?name=files/rooms/30f11559ba40ec4fc013/c48cb6746fe04d40064a3767d3033fec6d6777c9

POST https://firebasestorage.googleapis.com/v0/b//o?name=files%2Frooms%2F30f11559ba40ec4fc013%2Fc48cb6746fe04d40064a3767d3033fec6d6777c9
```

If I can be of any help with any of these issues please let me know 😎

**Comment by PatWie** [comment]
I currently lack a clear understanding of which stubs or functionality we need to re-implement from the Firebase API. This could potentially be a significant time-sink. I still prefer not to modify any frontend code and instead focus on providing correct request/response handling for all fetches. My goal is to eventually enable easy self-hosting on AWS Fargate. However, I must admit, I am not super keen in extensive programming during my spare time at the moment.

**Comment by codiflow** [comment]
> I still prefer not to modify any frontend code and instead focus on providing correct request/response handling for all fetches.

Totally agree with that. Unfortunately I never used Firebase before so I can't provide any help either. Maybe I get a better understanding while diving deeper into the code.

What I can say is that there is a working (but outdated) fork out there where the upstream code was modified so a httpStorage backend can be used instead of Firebase. This is the repository: https://github.com/squatica/excalidraw

> My goal is to eventually enable easy self-hosting on AWS Fargate.

OK I won't use any Amazon services so this is where I'm out 😉 Anyway – I'm sure your approach is so great that we can evolve it that it's an alternative to the current forks. Maybe I can find some other Go contributors who can help here 🙏

> The backend storage issue, i don't get. It works for me with link sharing using S3.

Could you maybe try if the backend also works without S3? I'm using an sqlite db at the moment and will try it with the directory storage today.

> Interestingly, I didn't know that one can embed arbitrary images 🙈. That's news to me.

Yes, that's one of the greatest features apart from already being a great piece of software at all 😎

**Comment by nhatsangvn** [comment]
Hey @PatWie ,

This is great work, I really appreciated. 

One issue that I got is even I set  (before starting the `./excaldraw-complete`)
```
export STORAGE_TYPE=sqlite
export DATA_SOURCE_NAME=/tmp/test.sqlite
```
the app seems doesn't use it. Although I can see `/tmp/test.sqlite` but when I draw some charts, that files doesn't change in size. Anyway that I can verify if the app actually use my sqlite db?

**Comment by PatWie** [comment]
I added a commit with logging (and updated README.md). I verified S3, filesystem, memory and sqlite storage works here on Linux.

**Comment by nhatsangvn** [comment]
Thank you @PatWie , I just re-tested it but the issue is still happened on my end.

When I started the server with
```
STORAGE_TYPE=sqlite DATA_SOURCE_NAME=test.db go run main.go --loglevel debug --listen=":3002"
```
I can see the `test.db` is there, but no new content is written.
<img width="1020" alt="image" src="https://github.com/PatWie/excalidraw-complete/assets/20770118/bf37c83c-1c90-4c1e-beec-a8f2eae4fbad">

After login to my domain (the app behind the reverse_proxy), open a live-collab, wait a few seconds and this issue happened.
<img width="688" alt="image" src="https://github.com/PatWie/excalidraw-complete/assets/20770118/1cada1d3-1709-4332-8516-2aa32cd19063">

**Comment by PatWie** [comment]
The issue in the screenshot you see, is probably the required SSL for firebase in the Front-ende implementation. One work-around is running excalidraw locally (Reverse-Proxy style) behind caddy, which does some magic to get valid self-signed, (and trusted) SSL certificates automatically into the browser without hassle.

**Comment by nhatsangvn** [comment]
Hi @PatWie , one new issue that I observed in the Room Server is: 
<img width="1177" alt="image" src="https://github.com/PatWie/excalidraw-complete/assets/20770118/7569e046-f4b6-4675-b509-2a576e85c15f">

when I try to open a colab room, sometimes the people in the rooms keep increasing although only 2 of my browsers are active.
![image](https://github.com/PatWie/excalidraw-complete/assets/20770118/2e57e3b8-e852-44c8-9f6f-0621a24b55d4)

Basically when I draw for sometimes (10s) and the number of people increases.

**Comment by PatWie** [comment]
This is unfortunately a known issue without a known root-cause:
https://github.com/PatWie/excalidraw-complete/issues/3
